### DISCUSSION TITLE
How to show errors in vs-code?

### DISCUSSION THREAD

**Discussion by njezersek** [nonara/ts-patch#79] [Q&A] [answered]
Hi everyone!
I just started using this tool a few weeks ago. Got something working but now I'm stuck. I hope someone here might be able to help me out.

## Background
I have created a transformer that enables operator overloading by replacing operators (`+`, `-`, `*`, `/`, ...) with function calls (`__add__`, `__sub__`, ...). When compiling the code with `tsc`, the errors are thrown as expected. For example `a + b` is transformed into `a.__add__(b)` and if `b` is the wrong type for `__add__` an error is thrown.

More details are available in this repo: https://github.com/njezersek/Adding-operator-overloading-to-TypeScript

## The problem
Everything works fine up to this point. But as the title suggests I can't figure out how to show correct errors in the editor. 

Firstly I thought I could get the language server working on the transformed code (_please correct me if this is not possible_).

So I tried to set `"typescript.tsdk": "./node_modules/typescript/lib"` in VS-Code to use the local patched version of TypeScript as noted in this comment - https://github.com/nonara/ts-patch/discussions/48#discussioncomment-971444. But to no avail.

Furthermore, I tried to patch the `tsserver.js` as described here - https://github.com/nonara/ts-patch/discussions/40#discussioncomment-977445. Sadly this caused the language server to crash every time. (TypeScript version: 4.7.2)

I also heard about writing a custom Langauge Server plugin. If that is necessary I am curious what would be the most straightforward  way to get the desired result, or if this is even possible (documentation for writing a Langage Service Plugin is very discouraging https://github.com/microsoft/TypeScript/wiki/Writing-a-Language-Service-Plugin)

**Comment by nonara** [comment, marked as answer]
Hello.

Unfortunately, patching language server does not work. I am going to deprecate it in future versions.

That said, you don't need a patch to run language server plugins. Those are natively supported in the `plugins` field.

Regarding showing errors, you can show errors for pre-transformed code, as it has something to report on. Reporting errors in the IDE on transformed code wouldn't be possible, because there is nothing to highlight.

I'm assuming you mean you want to report custom errors for code — this you would do with a language server plugin. If you google around, you should be able to find some examples. I know of several people who have done it; unfortunately, I don't have an exact repo to give as an example, but they are out there. People have added custom errors as well as filtered others out. Good luck!

**Reply by robert-w-gries** [reply]
> Unfortunately, patching language server does not work. I am going to deprecate it in future versions.

I have a use case for patching the language server as I am using this tool to write plugins to generate declaration files before the type checking phase. This doesn't break the language server error display and works decently well in my experience.

I know it's an odd use case for this tool, but I found it's the only way to properly extend the typescript compiler/language server. The built-in plugins for typescript don't allow me to do what I want to do.

**Reply by nonara** [reply]
@robert-w-gries Interesting! So long as there is a valid use, we'll keep it. Maybe we'll just add a warning message. Thank you for letting me know.
